crypto/evp/digest.c

```diff
@@ -113,56 +90,15 @@ int EVP_DigestInit_ex(EVP_MD_CTX *ctx, const EVP_MD *type, ENGINE *impl)
     if (type != NULL)
         ctx->reqdigest = type;
 
-    /* TODO(3.0): Legacy work around code below. Remove this */
-#if !defined(OPENSSL_NO_ENGINE) && !defined(FIPS_MODE)
-    /*
-     * Whether it's nice or not, "Inits" can be used on "Final"'d contexts so
-     * this context may already have an ENGINE! Try to avoid releasing the
-     * previous handle, re-querying for an ENGINE, and having a
-     * reinitialisation, when it may all be unnecessary.
-     */
-    if (ctx->engine && ctx->digest &&
-        (type == NULL || (type->type == ctx->digest->type)))
-        goto skip_to_init;
+#if !defined(FIPS_MODE)
+    {
+        int ret = 0;
 
-    if (type != NULL) {
-        /*
-         * Ensure an ENGINE left lying around from last time is cleared (the
-         * previous check attempted to avoid this if the same ENGINE and
-         * EVP_MD could be used).
-         */
-        ENGINE_finish(ctx->engine);
-        ctx->engine = NULL;
+        if (legacy_evp_digest_init_ex(ctx, type, impl, &ret))
```

Comment:
I looked at `legacy_evp_digest_init_ex`, and noticed that it does look at fetched stuff, and that might be overdoing stuff a bit.  So it seems to me that instead of a "do everything" function, it might be better with two, kind of "prepare" and "execute", so to say.  In code, I would expect to something along the lines of:

``` C
        if ((type = legacy_evp_digest_get_impl(ctx, type, impl)) != NULL)
            return legacy_evp_digest_init_ex(ctx, type);
```

`legacy_evp_digest_get_impl()` would then be able to get an engine implementation if there is such a thing, for example from an engine, and if none is available, it would simply return `type`.  `legacy_evp_digest_init_ex()` would then execute the initialization itself.

Reply:
Note that with that kind of division, there's also no need for `&ret`...

Reply:
That doesnt really work for legacy_evp_digest_init_ex() since there is a whole skip_to_init: branch that doesnt use the type.

Reply:
Yeah I know, it's not the easiest code to form as we wish, and I'm not saying that for the lack of trying.

I'm way too tired to be constructive right now, so I'll get tomorrow, yeah?